File: controllers/client_controller.py
from typing import Dict, List, Optional
from utils.logger_utils import log_error, log_info
from sqlalchemy.exc import IntegrityError
from fastapi import HTTPException
from datetime import datetime
from schemas.client_schema import ClientCreate, ClientUpdate
from repository.client_repository import ClientRepository

class ClientController:
    @staticmethod
    async def create_client(client_data: Dict, user_id: int) -> Dict:
        """
        Create a new client in the database.
        """
        log_info(f"Creating client with data: {client_data}")
        try:
            # Add user ID to client data
            client_data["createdBy"] = user_id
            
            # Validate and create client model
            client_create = ClientCreate(**client_data)
            
            # Use repository to create client
            result = await ClientRepository.create(client_create)
            
            if "error" in result:
                log_error(f"Error creating client: {result['error']}")
                raise HTTPException(
                    status_code=result.get("status", 500), 
                    detail=result["error"]
                )
            
            log_info(f"Client created successfully with ID: {result['client_id']}")
            return result
            
        except HTTPException:
            raise
        except Exception as e:
            log_error(f"Unexpected error creating client: {str(e)}")
            raise HTTPException(status_code=500, detail="Internal server error.")

    @staticmethod
    async def get_clients(
        page: int, 
        page_size: int = 10, 
        sort_by: Optional[str] = None, 
        sort_direction: Optional[str] = "asc", 
        keyword: Optional[str] = None
    ) -> Dict:
        """
        Retrieve a list of clients from the database.
        """
        if page < 1:
            raise HTTPException(status_code=400, detail="Page number must be greater than 0")
        
        log_info(f"Retrieving clients with page={page}, page_size={page_size}, keyword={keyword}")

        try:
            clients = await ClientRepository.get_paginated(
                page=page,
                page_size=page_size,
                sort_by=sort_by,
                sort_direction=sort_direction,
                keyword=keyword
            )
            
            return clients
            
        except Exception as e:
            log_error(f"Error fetching clients: {str(e)}")
            raise HTTPException(status_code=500, detail="Internal server error.")

    @staticmethod
    async def get_client_by_id(client_id: int) -> Dict:
        """
        Retrieve a client by its ID.
        """
        log_info(f"Fetching client with ID: {client_id}")
        try:
            client = await ClientRepository.get_by_id(client_id)
            
            if not client:
                raise HTTPException(status_code=404, detail="Client not found")
                
            return client.model_dump()
            
        except HTTPException:
            raise
        except Exception as e:
            log_error(f"Error fetching client {client_id}: {str(e)}")
            raise HTTPException(status_code=500, detail="Internal server error.")

    @staticmethod
    async def validate_client_exists(client_id: int) -> Dict:
        """
        Validate that a client exists and return its data.
        """
        try:
            client = await ClientController.get_client_by_id(client_id)
            return client
        except HTTPException as e:
            log_error(f"Client with ID {client_id} not found")
            raise e

    @staticmethod
    async def update_client(client_id: int, client_data: Dict, user_id: Optional[int] = None) -> Dict:
        """
        Update a client's data by its ID.
        """
        log_info(f"Updating client {client_id} with data: {client_data}")
        try:
            # Check if client exists
            await ClientController.validate_client_exists(client_id)
            
            # Add updatedBy user ID if provided
            if user_id:
                client_data["updatedBy"] = user_id
            
            # Validate update data
            client_update = ClientUpdate(**client_data)
            
            # Update client
            result = await ClientRepository.update(client_id, client_update)
            
            if "error" in result:
                log_error(f"Error updating client: {result['error']}")
                raise HTTPException(
                    status_code=result.get("status", 500), 
                    detail=result["error"]
                )

            log_info(f"Client {client_id} updated successfully")
            return result
            
        except HTTPException:
            raise
        except Exception as e:
            log_error(f"Error updating client {client_id}: {str(e)}")
            raise HTTPException(status_code=500, detail="Internal server error.")

    @staticmethod
    async def delete_client(client_id: int, user_id: Optional[int] = None) -> Dict:
        """
        Soft delete a client by its ID.
        """
        log_info(f"Deleting client with ID: {client_id}")
        try:
            # Check if client exists
            await ClientController.validate_client_exists(client_id)
            
            # Soft delete client
            result = await ClientRepository.soft_delete(client_id, user_id or 0)
            
            if "error" in result:
                log_error(f"Error deleting client: {result['error']}")
                raise HTTPException(
                    status_code=result.get("status", 500), 
                    detail=result["error"]
                )

            log_info(f"Client {client_id} deleted successfully")
            return result
            
        except HTTPException:
            raise
        except Exception as e:
            log_error(f"Error deleting client {client_id}: {str(e)}")
            raise HTTPException(status_code=500, detail="Internal server error.")

    @staticmethod
    async def search_clients(keyword: str) -> List[Dict]:
        """
        Search clients by keyword.
        """
        log_info(f"Searching clients with keyword: {keyword}")
        try:
            clients = await ClientRepository.search_by_keyword(keyword)
            return [client.model_dump() for client in clients]
            
        except Exception as e:
            log_error(f"Error searching clients: {str(e)}")
            raise HTTPException(status_code=500, detail="Internal server error.")

File: src/controllers/client.controller.ts
import { ClientRepository } from "../repository/client"
import { UserRepository } from "../repository/user"
import { logError, logInfo } from "../utils/logger"
import { paginationMeta } from "../utils/pagination"
import { ClientModel } from "../models/client.model"
import { UserModel } from "../models/user.model"
import type { IClientCreate, IClientOutput, IClientUpdate, IClientView } from "../interfaces/client.interfaces"
import type { IUserView } from "../interfaces/user.interfaces"

type ControllerResult<T> = { kind: 'success'; data: T } | { kind: 'error'; error: string; status: number }

async function buildUserView(userId: number | null | undefined): Promise<IUserView> {
  if (!userId) return UserModel.toView(null)
  const user = await UserRepository.getById(userId)
  return UserModel.toView(user)
}

export class ClientController {
  static async create(data: IClientCreate, userId: number): Promise<ControllerResult<IClientOutput>> {
    try {
      const payload = ClientModel.createPayload(data)
      const result = await ClientRepository.create({ ...payload, createdBy: userId, createdAt: new Date() })
      const userView = await buildUserView(userId)
      logInfo(`Client created: ${result.id}`)
      return { kind: 'success', data: ClientModel.toOutput(result, userView, userView) }
    } catch (e) {
      logError(`Error creating client: ${e}`)
      return { kind: 'error', error: "Failed to create client", status: 500 }
    }
  }

  static async getAll(page: number, pageSize: number, keyword?: string, sortBy?: string, sortDir?: string): Promise<ControllerResult<{ data: IClientView[]; meta: any }>> {
    if (page < 1) return { kind: 'error', error: "Page must be >= 1", status: 400 }
    try {
      const { data, total } = await ClientRepository.getAll(page, pageSize, keyword, sortBy, sortDir)
      const transformedData = data.map(ClientModel.toView)
      return { kind: 'success', data: { data: transformedData, meta: paginationMeta(total, page, pageSize) } }
    } catch (e) {
      logError(`Error fetching clients: ${e}`)
      return { kind: 'error', error: "Failed to fetch clients", status: 500 }
    }
  }

  static async getById(id: number): Promise<ControllerResult<IClientOutput>> {
    try {
      const client = await ClientRepository.getById(id)
      if (!client) return { kind: 'error', error: "Client not found", status: 404 }
      const [createdBy, updatedBy] = await Promise.all([
        buildUserView(client.createdBy),
        buildUserView(client.updatedBy),
      ])
      return { kind: 'success', data: ClientModel.toOutput(client, createdBy, updatedBy) }
    } catch (e) {
      logError(`Error fetching client ${id}: ${e}`)
      return { kind: 'error', error: "Failed to fetch client", status: 500 }
    }
  }

  static async update(id: number, data: IClientUpdate, userId: number): Promise<ControllerResult<IClientOutput>> {
    try {
      const existing = await ClientRepository.getById(id)
      if (!existing) return { kind: 'error', error: "Client not found", status: 404 }

      const payload = ClientModel.updatePayload(data)
      const result = await ClientRepository.update(id, { ...payload, updatedBy: userId, updatedAt: new Date() })
      const [createdBy, updatedBy] = await Promise.all([
        buildUserView(existing.createdBy),
        buildUserView(userId),
      ])
      logInfo(`Client updated: ${id}`)
      return { kind: 'success', data: ClientModel.toOutput(result, createdBy, updatedBy) }
    } catch (e) {
      logError(`Error updating client ${id}: ${e}`)
      return { kind: 'error', error: "Failed to update client", status: 500 }
    }
  }

  static async search(keyword: string): Promise<ControllerResult<IClientView[]>> {
    try {
      const data = await ClientRepository.search(keyword)
      const transformedData = data.map(ClientModel.toView)
      return { kind: 'success', data: transformedData }
    } catch (e) {
      logError(`Error searching clients: ${e}`)
      return { kind: 'error', error: "Failed to search clients", status: 500 }
    }
  }

  static async delete(id: number, userId: number): Promise<ControllerResult<{ message: string }>> {
    try {
      const existing = await ClientRepository.getById(id)
      if (!existing) return { kind: 'error', error: "Client not found", status: 404 }

      await ClientRepository.softDelete(id, userId)
      logInfo(`Client deleted: ${id}`)
      return { kind: 'success', data: { message: "Client deleted" } }
    } catch (e) {
      logError(`Error deleting client ${id}: ${e}`)
      return { kind: 'error', error: "Failed to delete client", status: 500 }
    }
  }
}